flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -0,0 +1,248 @@
+			ExecutionVertexID executionVertexId,
+			Collection<TaskManagerLocation> preferredLocationsBasedOnState,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+
+		if (!preferredLocationsBasedOnState.isEmpty()) {
+			return CompletableFuture.completedFuture(preferredLocationsBasedOnState);
+		}
+
+		return getPreferredLocationsBasedOnInputs(executionVertexId, inputsLocationsRetriever);
+	}
+
+	/**
+	 * Gets the location preferences of the execution, as determined by the locations
+	 * of the predecessors from which it receives input data.
+	 * If there are more than {@link ExecutionVertex#MAX_DISTINCT_LOCATIONS_TO_CONSIDER} different locations of source data,
+	 * or neither the sources have not been started nor will be started with the execution together,
+	 * this method returns an empty collection to indicate no location preference.
+	 *
+	 * @return The preferred locations based in input streams, or an empty iterable,
+	 *         if there is no input-based preference.
+	 */
+	@VisibleForTesting
+	static CompletableFuture<Collection<TaskManagerLocation>> getPreferredLocationsBasedOnInputs(
+			ExecutionVertexID executionVertexId,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+		CompletableFuture<Collection<TaskManagerLocation>> preferredLocations =
+				CompletableFuture.completedFuture(Collections.emptyList());
+
+		Collection<CompletableFuture<TaskManagerLocation>> locationsFutures = new ArrayList<>();
+
+		Collection<Collection<ExecutionVertexID>> allProducers =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(executionVertexId);
+		for (Collection<ExecutionVertexID> producers : allProducers) {
+
+			for (ExecutionVertexID producer : producers) {
+				Optional<CompletableFuture<TaskManagerLocation>> optionalLocationFuture =
+						inputsLocationsRetriever.getTaskManagerLocation(producer);
+				optionalLocationFuture.ifPresent(
+						(locationFuture) -> locationsFutures.add(locationFuture));
+				// If the parallelism is large, wait for all futures coming back may cost a long time.
+				if (locationsFutures.size() > MAX_DISTINCT_LOCATIONS_TO_CONSIDER) {
+					locationsFutures.clear();
+					break;
+				}
+			}
+
+			CompletableFuture<Collection<TaskManagerLocation>> uniqueLocationsFuture =
+					FutureUtils.combineAll(locationsFutures).thenApply(
+							(locations) -> {
+								Set<TaskManagerLocation> uniqueLocations = new HashSet<>(locations);
+								return uniqueLocations;
+							});
+			preferredLocations = preferredLocations.thenCombine(
+					uniqueLocationsFuture,
+					(locationsOnOneEdge, locationsOnAnotherEdge) -> {
+						if ((!locationsOnOneEdge.isEmpty() && locationsOnAnotherEdge.size() > locationsOnOneEdge.size())
+								|| locationsOnAnotherEdge.isEmpty()) {
+							return locationsOnOneEdge;
+						} else {
+							return locationsOnAnotherEdge;
+						}
+					});
+			locationsFutures.clear();
```

Comment:
why use `clear` rather than `new`, in most cases `new` is more efficient. (https://stackoverflow.com/questions/6961356/list-clear-vs-list-new-arraylistinteger)

Reply:
Probably because `.clear()` is closer to the legacy code. Can be changed later.